### ISSUE TITLE
Generalize `uploadResults()` for use in other packages

### ISSUE THREAD

**Issue by msuchard** [OHDSI/CohortDiagnostics#442]
I'd like to propose that we generalize the signature for `uploadResults()` so that this function can be (re-)used in other packages.   The proposal is to change this signature to:

```
uploadResults <- function(connectionDetails = NULL,
                          schema,
                          zipFileName,
                          forceOverWriteOfSpecifications = FALSE,
                          purgeSiteDataBeforeUploading = TRUE,
                          tempFolder = tempdir(),
                          specification = getResultsDataModelSpecifications()) # CHANGE HERE
```

then other packages (like `LegendT2dm` could provide their own `specification` arguments.

**Comment by msuchard** [comment]
If you (@schuemie @gowthamrao) agree, I'll make the changes.  Let me know.

**Comment by msuchard** [comment]
or move this function (and its dependencies) into, say, `OhdsiRTools` and then write:

```
CohortDiagnostics::uploadResults <- function(connectionDetails,  # NOTE: THIS CANNOT BE NULL (CURRENT DEFAULT)
                          schema,
                          zipFileName,
                          forceOverWriteOfSpecifications = FALSE,
                          purgeSiteDataBeforeUploading = TRUE,
                          tempFolder = tempdir()) {

OhdsiRTools::uploadResults(connectionDetails, 
                          schema,
                          zipFileName,
                          forceOverWriteOfSpecifications,
                          purgeSiteDataBeforeUploading,
                          tempFolder,
                          specification = getResultsDataModelSpecifications()) 
}
```

**Comment by gowthamrao** [comment]
i love it! OhdsiRTools sounds like a great new home. 

- For the specification - are you thinking of requiring a DDL script like https://github.com/OHDSI/CohortDiagnostics/blob/master/inst/sql/postgresql/CreateResultsDataModel.sql or you plan to generate the DDL from `getResultsDataModelSpecifications()`. 
- Are you planning to only support postgres or do you want to support translations (we have an informal position of only supporting postgres for results, because we found shiny to work well with postgres and the use of 'pool' connection that is only supported by postgres i believe). Plus postgres is free
- i also propose to add a lot of checks using the specifications on `getResultsDataModelSpecifications()`. Martijn had written a lot of data integrity checks in cohort diagnostics by comparing to resultsDataModelSpecifications


We can target cohort diagnostics 2.2 if you like

**Comment by schuemie** [comment]
I agree that our approach to uploading the results is becoming a repeating pattern that warrants extraction to a reusable function. One thing to keep in mind is that the function does need to know what the primary key field(s) of a table are. So we should make it a hard requirement that folks provide a results model specification like [this one](https://github.com/ohdsi-studies/Eumaeus/blob/main/inst/settings/resultsModelSpecs.csv). The advantage is that these specs can be used both to create the table structure, and to generate the data model documentation.

I'm not sure about OhdsiRTools as its new home though. Right now OhdsiRTools is a package only for package developers, not package users. This is also the reason why OhdsiRTools is currently not in HADES. Maybe we could create a new OhdsiStudies package? (Or OhdsiStudyPackages?)

I propose to stick to Postgres for now. Uploading large amounts of data to existing tables is highly optimized in DatabaseConnector for Postgres , but not so much for other platforms (with the possible exception of RedShift when bulk uploading is configured).

**Comment by schuemie** [comment]
(An alternative destination might be OhdsiSharing)

**Comment by leeevans** [comment]
I agree that OhdsiSharing is a good home for this function and I'd be happy to start working on implementing it there.

**Comment by leeevans** [comment]
Should we also add an (OHDSI network) siteId parameter? I believe it could be useful in a couple ways:

a) Use it to create a new database schema for the site - if we have a use case for creating separate site schemas?
b) Use it to populate a site id column in the generated results tables -  downstream processes can use that column to merge the data across sites into a single combined table. 

The OHDSI cloud admin can assign a unique OHDSI network site id along with providing the database userid/password credentials to the new site user responsible for uploading the results.

If we do want to assign a unique OHDSI network site id, we could perhaps use the same approach used for assigning java namespaces?

https://en.wikipedia.org/wiki/Reverse_domain_name_notation

**Comment by msuchard** [comment]
I'm certainly in favor of `OhdsiSharing` as a good home for `uploadResults`.  My suggestion would be to _first_ port over the existing functionality (to remove lots of code duplication in many places) and _then_ add additional functionality *while* maintaining backwards compatibility.

`uploadResults` is currently agnostic to table columns (a table can have a `siteId` or not depending on its `specification`).  

A function to create a new database schema would (and currently is, e.g. `CohortDiagnostics::createResults...`) separate from the `uploading` functionality.

**Comment by gowthamrao** [comment]
@leeevans  can we transfer the issue from Cohort Diagonstics to OhdsiSharing?

Once its added to OhdsiSharing, i can remove it from Cohort Diagnostics
